controllers/metal3.io/hostfirmwaresettings_controller.go

```diff
@@ -0,0 +1,448 @@
+		reqLogger.Info("provisioner is not ready", "RequeueAfter:", provisionerNotReadyRetryDelay)
+		return ctrl.Result{Requeue: true, RequeueAfter: provisionerNotReadyRetryDelay}, nil
+	}
+
+	// Get the data from Ironic and update HFS, the call to provisioner may fail if settings can't
+	// be retrieved for the bios_interface. Just log it and don't retry
+	if err = r.updateHostFirmwareSettings(prov, info); err != nil {
+		return ctrl.Result{}, errors.Wrap(err, "Could not update hostFirmwareSettings")
+	}
+
+	for _, e := range info.events {
+		r.publishEvent(req, e)
+	}
+
+	// only run again on changes
+	return ctrl.Result{}, nil
+}
+
+// Get the firmware settings from the provisioner and update hostFirmwareSettings
+func (r *HostFirmwareSettingsReconciler) updateHostFirmwareSettings(prov provisioner.Provisioner, info *rInfo) (err error) {
+
+	info.log.Info("retrieving firmware settings and saving to resource")
+
+	// Get the current settings and schema
+	currentSettings, schema, err := prov.GetFirmwareSettings(true)
+	if err != nil {
+		return errors.Wrap(err, "could not get firmware settings from provisioner")
+	}
+
+	// get or create a firmwareSchema to hold schema
+	firmwareSchema, err := r.getOrCreateFirmwareSchema(info, schema)
+	if err != nil {
+		return errors.Wrap(err, "could not get/create firmware schema")
+	}
+
+	// Set hostFirmwareSetting to use this schema
+	info.hfs.Status.FirmwareSchema = &metal3v1alpha1.SchemaReference{
+		Namespace: firmwareSchema.ObjectMeta.Namespace,
+		Name:      firmwareSchema.ObjectMeta.Name}
+
+	if err = r.updateSpecAndStatus(info, currentSettings, firmwareSchema); err != nil {
+		return errors.Wrap(err, "could not update hostFirmwareSettings")
+	}
+
+	return nil
+}
+
+// Update the HostFirmwareSettings resource using the settings and schema from provisioner
+func (r *HostFirmwareSettingsReconciler) updateSpecAndStatus(info *rInfo, settings metal3v1alpha1.SettingsMap, schema *metal3v1alpha1.FirmwareSchema) (err error) {
+
+	if info.hfs.Status.Settings == nil {
+		info.hfs.Status.Settings = make(metal3v1alpha1.SettingsMap)
+	}
+
+	// Update Spec and Status
+	specDirty := false
+	for k, v := range settings {
+		// Some vendors include encrypted password fields, don't add these
+		if strings.Contains(k, "Password") {
+			continue
+		}
+		info.hfs.Status.Settings[k] = v
+
+		// Don't include setting in Spec if it cannot be changed (ReadOnly is set)
+		// or if it unique to this particular host, in order to prevent the unique
+		// settings from being copied to other hosts
+		if schema != nil {
+			if schemaSetting, ok := schema.Spec.Schema[k]; ok {
+				if (schemaSetting.ReadOnly != nil && *schemaSetting.ReadOnly == true) ||
+					(schemaSetting.Unique != nil && *schemaSetting.Unique == true) {
+					continue
+				}
+			}
+		}
+		// Copy setting to Spec only if doesn't exist as setting may have been set by user
+		if _, ok := info.hfs.Spec.Settings[k]; !ok {
+			info.hfs.Spec.Settings[k] = intstr.FromString(v)
```

Comment:
I must admit, I don't remember the design too well, but updating the spec seems confusing to me. Does it mean that if a setting gets changed by a 3rd party (from within the machine or even automatically by the BMC), we'll try to modify it because it no longer matches the cached spec?

I would expect the current settings to go to Status, while Spec only containing what the user wants us to assert.

Reply:
Note that the Status only gets copied to the Spec if that entry doesn't exist in the Spec. So the way it will work is first time we retrieve the data from Ironic both the Status and the Spec will be updated to the same values. This improves the user experience in that the actual value in the Spec can be just changed when desired without having to input the setting name. 

So your question is if the STATUS field gets updated by a 3rd party?  I would think it that case the reconciler would run and update the Status to the value from Ironic, in that case there would not be a mismatch with the Spec. The only time we'll try to modify the entry is if the Spec field actually gets updated

Reply:
No, I mean what if a setting is updated in the machine by something? E.g. some hardware have boot device as BIOS setting, ironic does update that. Will it cause any actions?

Reply:
> No, I mean what if a setting is updated in the machine by something? E.g. some hardware have boot device as BIOS setting, ironic does update that. Will it cause any actions?

Yeah, good point, I can see that if the actual setting gets updated outside of Ironic it can cause a an inadvertent update to the old value. I've removed the syncing of the Spec on update. This will add an extra step for the user to create the entry in the Spec when making a change but it will prevent this spurious update.